### ISSUE TITLE
session handling

### ISSUE THREAD

**Issue by glyph** [twisted/klein#134]
Speaking as its author: Twisted's session API is irredeemable garbage, and Klein realistically can't use it.  We should have a session-management API.

**Comment by glyph** [comment]
#126 depends on this

**Comment by notoriousno** [comment]
Is this a Twisted issue or Klein issue?  Most web frameworks (tornado and expressjs come to mind) don't provide session handling and it's explicitly stated in their docs.  A session API would be beneficial but I'm just worried about the scope of it all.  Session management is highly opinionated and a bit of a rabbit hole for devs.  Trying to find a generic means to handle sessions might lead to more headaches than it's worth.  I would argue something like middleware functionally would be a better way to combat sessions.

PS
What's wrong with the current session framework in Twisted?  It seems to be in line with what others do.  FYI I only looked at the session code very briefly...a long time ago...and just now...after a long time.

* [ExpressJS Middleware ](https://expressjs.com/en/guide/writing-middleware.html)

**Comment by glyph** [comment]
@notoriousno - The main thing is that thanks to its synchronous session-store API and hard-coded ID generation, it effectively thinks that your sessions are all going to be local state in a single process.  I'm not even sure if it's *possible* to work around that.

The reason we need session functionality is #126; you can't securely handle *any* input from a browser unless you have a mechanism for identifying actors and associating tokens with their intent, which means server-side state, which means session IDs.  You can see my prototype branch for this at https://github.com/glyph/klein/tree/form-hacks if you want to opine :).  Hopefully I'll do a proper PR soon, once I'm sure a whole application hangs together nicely.

I haven't had a chance to look at that particular link, I'm not really a fan of "middleware" as a concept.  The problem is that it takes a bunch of very specific, almost always security-sensitive attributes that an application needs, and attempts to factor them out into  a separate layer, so you can run the same code with or without sessions, or with or without authentication.  Generally speaking I don't think code which requires parameters should have the option to run *without* those parameters, so if you need "a logged in user" or "a session ID", you should very tightly bind to a system that is going to provide those to you.

**Comment by notoriousno** [comment]
Finally had some time over the long weekend to take a look at the session branch. It looks an awful lot like the ``cred`` framework. Can that not be used instead of making a new framework? I vaguely recall ``cred`` has drawbacks when it comes to HTTP servers so that might be the reason. I failed miserably last time I tried to use ``cred`` anyways :D.

Other than that, I think I understand how it's supposed to work and will be trying to implement something simple today with your session module. It was a bit hard to follow with all the new plating stuff

**Comment by glyph** [comment]
@notoriousno what similarities do you see to `cred`?  This is more like the bottom half of the glue that would allow you to hook up `cred` to HTTP.

**Comment by notoriousno** [comment]
My mistake it didn't dawn on me until now that the database portion is not implemented by Twisted.

**Comment by IlyaSkriblovsky** [comment]
IMHO sessions mechanism is not required for CSRF protection. You can set csrf cookie with random value when GETting the <form> and also inject `<input type="hidden" value="[csrf-value]"/>` into it. Then just compare POSTed value with the cookie from the request.

**Comment by glyph** [comment]
@IlyaSkriblovsky "sessions" is a highly generic term here; the thing you're describing is just a client-only session store with no server-side state.  The existing thing I've implemented in my form-hacks branch can facilitate that kind of storage, and could either just compare cookies to form fields, store stuff in a backend, or use a signing mechanism like itsdangerous.  What this ticket is about is having an abstract interface for that so form processing isn't tied to one particular implementation.

**Comment by glyph** [comment]
Fixed in #246 .
